# Analyzing Used Car Listings on eBay Kleinanzeigen

We will be working on a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The dataset was originally scraped and uploaded to Kaggle. The version of the dataset we are working with is a sample of 50,000 data points that was prepared by Dataquest including simulating a less-cleaned version of the data.

The data dictionary provided with data is as follows:

* dateCrawled - When this ad was first crawled. All field-values are taken from this date.
* name - Name of the car.
* seller - Whether the seller is private or a dealer.
* offerType - The type of listing
* price - The price on the ad to sell the car.
* abtest - Whether the listing is included in an A/B test.
* vehicleType - The vehicle Type.
* yearOfRegistration - The year in which which year the car was first registered.
* gearbox - The transmission type.
* powerPS - The power of the car in PS.
* model - The car model name.
* kilometer - How many kilometers the car has driven.
* monthOfRegistration - The month in which which year the car was first registered.
* fuelType - What type of fuel the car uses.
* brand - The brand of the car.
* notRepairedDamage - If the car has a damage which is not yet repaired.
* dateCreated - The date on which the eBay listing was created.
* nrOfPictures - The number of pictures in the ad.
* postalCode - The postal code for the location of the vehicle.
* lastSeenOnline - When the crawler saw this ad last online.

The aim of this project is to clean the data and analyze the included used car listings.

In [1]:
import pandas as pd
import numpy as np
#from csv import reader
autos = pd .read_csv('autos.csv', encoding = 'Latin-1')
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Our dataset contains 20 columns, most of which are stored as strings. There are a few columns with null values, but no columns have more than ~20% null values. There are some columns that contain dates stored as strings.

We'll start by cleaning the column names to make the data easier to work with.

# Clean Columns

In [2]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

We'll make a few changes here:

* Change the columns from camelcase to snakecase.
* Change a few wordings to more accurately describe the columns.

In [3]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

# Initial Data Exploration and Cleaning
We'll start by exploring the data to find obvious areas where we can clean the data.

In [4]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-16 21:50:53  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

Our initial observations:

* There are a number of text columns where all (or nearly all) of the values are the same:
    * seller
    * offer_type
* The num_photos column looks odd, we'll need to investigate this further.

In [5]:
autos['num_photos'].value_counts()

0    50000
Name: num_photos, dtype: int64

It looks like the num_photos column has 0 for every column. We'll drop this column, plus the other two we noted as mostly one value.

In [6]:
autos = autos.drop(['num_photos', 'seller', 'offer_type'], axis=1)

In [7]:
autos['price'] = (autos['price']
                          .str.replace('$','')
                          .str.replace(',','')
                          .astype(int)
                 )
autos["price"].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [8]:
#autos['odometer'].head()
autos['odometer'] = (autos['odometer']
                         .str.replace(',','')
                         .str.replace('km','')
                         .astype(int)
                    )
autos.rename({'odometer': 'odometer_km'}, axis=1, inplace=True)
autos['odometer_km'].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

# Exploring Odometer and Price

In [9]:
print(autos['odometer_km'].unique().shape)
autos['odometer_km'].value_counts()

(13,)


150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

We can see that the values in this field are rounded, which might indicate that sellers had to choose from pre-set options for this field. Additionally, there are more high mileage than low mileage vehicles.

In [10]:
#unique:重複を削除した配列を返す shape:配列の形状を把握
print(autos['price'].unique().shape)
print(autos['price'].describe())
autos['price'].value_counts().head(10)

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
Name: price, dtype: int64

Again, the prices in this column seem rounded, however given there are 2357 unique values in the column, that may just be people's tendency to round prices on the site.

There are 1,421 cars listed with $0 price - given that this is only 2% of the of the cars, we might consider removing these rows. The maximum price is one hundred million dollars, which seems a lot, let's look at the highest prices further.

In [11]:
autos['price'].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

In [12]:
autos['price'].value_counts().sort_index(ascending=True).head(20)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
20       4
25       5
29       1
30       7
35       1
Name: price, dtype: int64

There are a number of listings with prices below \$30, including about 1,500 at \$0. There are also a small number of listings with very high values, including 14 at around or over $1 million.

Given that eBay is an auction site, there could legitimately be items where the opening bid is \$1. We will keep the \$1 items, but remove anything above \$350,000, since it seems that prices increase steadily to that number and then jump up to less realistic numbers.

In [13]:
autos = autos[autos['price'].between(1,351000)]
autos['price'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

# Exploring the date columns
There are a number of columns with date information:

* date_crawled
* registration_month
* registration_year
* ad_created
* last_seen

These are a combination of dates that were crawled, and dates with meta-information from the crawler. The non-registration dates are stored as strings.

We'll explore each of these columns to learn more about the listings.

In [14]:
autos[['date_crawled', 'ad_created', 'last_seen']][:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [15]:
#sorted by index(date)
#series.str[:10] 最初の9番目の文字までスライスする
#normalize:出現頻度の合計が１となるように返す
#dropna:欠損値をカウントせずに飛ばす。Trueにするとカウントしません。初期値True
(autos['date_crawled']
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
)

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [16]:
#sorted by values
(autos['date_crawled']
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_values()
)

2016-04-07    0.001400
2016-04-06    0.003171
2016-03-18    0.012911
2016-04-05    0.013096
2016-03-06    0.014043
2016-03-13    0.015670
2016-03-05    0.025327
2016-03-24    0.029342
2016-03-16    0.029610
2016-03-27    0.031092
2016-03-25    0.031607
2016-03-17    0.031628
2016-03-31    0.031834
2016-03-10    0.032184
2016-03-26    0.032204
2016-03-23    0.032225
2016-03-11    0.032575
2016-03-22    0.032987
2016-03-09    0.033090
2016-03-08    0.033296
2016-04-01    0.033687
2016-03-30    0.033687
2016-03-29    0.034099
2016-03-15    0.034284
2016-03-19    0.034778
2016-03-28    0.034860
2016-04-02    0.035478
2016-03-07    0.036014
2016-04-04    0.036487
2016-03-14    0.036549
2016-03-12    0.036920
2016-03-21    0.037373
2016-03-20    0.037887
2016-04-03    0.038608
Name: date_crawled, dtype: float64

Looks like the site was crawled daily over roughly a one month period in March and April 2016. The distribution of listings crawled on each day is roughly uniform.

In [17]:
#いつ車が売れたかを暗示している
(autos['last_seen']
        .str[:10]
        .value_counts(normalize=True,dropna=False)
        .sort_index()
)

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

The crawler recorded the date it last saw any listing, which allows us to determine on what day a listing was removed, presumably because the car was sold.

The last three days contain a disproportionate amount of 'last seen' values. Given that these are 6-10x the values from the previous days, it's unlikely that there was a massive spike in sales, and more likely that these values are to do with the crawling period ending and don't indicate car sales.

In [18]:
(autos['ad_created']
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
)

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21 

There is a large variety of ad created dates. Most fall within 1-2 months of the listing date, but a few are quite old, with the oldest at around 9 months.

In [19]:
autos['registration_year'].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The year that the car was first registered will likely indicate the age of the car. Looking at this column, we note some odd values. The minimum value is 1000, long before cars were invented and the maximum is 9999, many years into the future.

# Dealing with Incorrect Registration Year Data
Because a car can't be first registered before the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate. Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.

One option is to remove the listings with these values. Let's determine what percentage of our data has invalid values in this column:

In [20]:
#shape　要素数を行列で返す shape[0]は(48565, )のindex[0]を返す
#~は選択dfではない方を返す
(~autos['registration_year'].between(1900,2016)).sum() / autos.shape[0]
#(autos["registration_year"].between(1900,2016)).sum() / autos.shape[0]

0.038793369710697

Given that this is less than 4% of our data, we will remove these rows.

In [21]:
autos['registration_year'].value_counts().sort_index(ascending=False).head(20)

9999       3
9000       1
8888       1
6200       1
5911       1
5000       4
4800       1
4500       1
4100       1
2800       1
2019       2
2018     470
2017    1392
2016    1220
2015     392
2014     663
2013     803
2012    1310
2011    1623
2010    1589
Name: registration_year, dtype: int64

In [22]:
autos['registration_year'].value_counts().sort_index().head(20)

1000    1
1001    1
1111    1
1800    2
1910    5
1927    1
1929    1
1931    1
1934    2
1937    4
1938    1
1939    1
1941    2
1943    1
1948    1
1950    3
1951    2
1952    1
1953    1
1954    2
Name: registration_year, dtype: int64

In [23]:
autos = autos[autos['registration_year'].between(1900,2016)]
autos['registration_year'].describe()

count    46681.000000
mean      2002.910756
std          7.185103
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

In [24]:
autos['registration_year'].value_counts(normalize=True, dropna=False).head(10)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
Name: registration_year, dtype: float64

It appears that most of the vehicles were first registered in the past 20 years.

# Exploring Price by Brand

In [25]:
print(autos['brand'].unique().shape)
autos['brand'].value_counts(normalize=True)

(40,)


volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

German manufacturers represent four out of the top five brands, almost 50% of the overall listings. Volkswagen is by far the most popular brand, with approximately double the cars for sale of the next two brands combined.

There are lots of brands that don't have a significant percentage of listings, so we will limit our analysis to brands representing more than 5% of total listings.

In [26]:
brand_counts = autos['brand'].value_counts(normalize=True)
common_brands = brand_counts[brand_counts > .05].index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [27]:
brand_mean_prices = {}
for brand in common_brands:
    brand_only = autos[autos['brand'] == brand]
    mean_price = brand_only['price'].mean()
    brand_mean_prices[brand] = int(mean_price)

brand_mean_prices

{'audi': 9336,
 'bmw': 8332,
 'ford': 3749,
 'mercedes_benz': 8628,
 'opel': 2975,
 'volkswagen': 5402}

Of the top 5 brands, there is a distinct price gap:

* Audi, BMW and Mercedes Benz are more expensive
* Ford and Opel are less expensive
* Volkswagen is in between - this may explain its popularity, it may be a 'best of 'both worlds' option.

# Exploring Mileage

In [28]:
bmp_series = pd.Series(brand_mean_prices)
pd.DataFrame(bmp_series, columns=['mean_price'])

mean_price
audi                 9336
bmw                  8332
ford                 3749
mercedes_benz        8628
opel                 2975
volkswagen           5402

In [29]:
brand_mean_mileage = {}
for brand in common_brands:
    brand_only = autos[autos['brand'] == brand]
    mean_mileage = brand_only['odometer_km'].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

bmm_series = pd.Series(brand_mean_mileage)
pd.DataFrame(bmm_series,columns=['mean_mileage'])

mean_mileage
audi                 129157
bmw                  132572
ford                 124266
mercedes_benz        130788
opel                 129310
volkswagen           128707

In [30]:
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)
mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)

In [31]:
brand_info = pd.DataFrame(mean_mileage, columns=['mean_mileage'])
brand_info

mean_mileage
bmw                  132572
mercedes_benz        130788
opel                 129310
audi                 129157
volkswagen           128707
ford                 124266

In [32]:
brand_info['mean_price'] = mean_prices
brand_info

mean_mileage  mean_price
bmw                  132572        8332
mercedes_benz        130788        8628
opel                 129310        2975
audi                 129157        9336
volkswagen           128707        5402
ford                 124266        3749

The range of car mileages does not vary as much as the prices do by brand, instead all falling within 10% for the top brands. There is a slight trend to the more expensive vehicles having higher mileage, with the less expensive vehicles having lower mileage.

# Next steps
* Data cleaning next steps:
    * Identify categorical data that uses german words, translate them and map the values to their english counterparts
    * Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321.
    * See if there are particular keywords in the name column that you can extract as new columns

* Analysis next steps:
    * Find the most common brand/model combinations
    * Split the odometer_km into groups, and use aggregation to see if average prices follows any patterns based on the milage.
    * How much cheaper are cars with damage than their non-damaged counterparts?

In [33]:
'''
german_dict = {
    'Ü' : 'U',
    'Ä' : 'A',
    'Ö' : 'O'
}
autos['name'] = (autos['name']
                        .str.replace('Ü','U')
                        .str.replace('Ä', 'A')
                        .str.replace('Ö', 'O')
                )
autos['name'].head()
'''

"\ngerman_dict = {\n    'Ü' : 'U',\n    'Ä' : 'A',\n    'Ö' : 'O'\n}\nautos['name'] = (autos['name']\n                        .str.replace('Ü','U')\n                        .str.replace('Ä', 'A')\n                        .str.replace('Ö', 'O')\n                )\nautos['name'].head()\n"

# Identify categorical data that uses german words, translate them and map the values to their english counterparts

In [34]:
#seller,gearbox,unrepaired_damage
#seller is dropped
#manuell:manual, automatik:automatic, nein:no, ja:yes

print(autos.columns)
german_dict_gear_type = {
    'manuell' : 'manual',
    'automatik' : 'automatic'
}
german_dict_ud = {
    'ja' : 'yes' ,
    'nein' : 'no' 
}
autos['gearbox'] = autos['gearbox'].map(german_dict_gear_type)
autos['unrepaired_damage'] = autos['unrepaired_damage'].map(german_dict_ud)
print(autos['gearbox'].unique())
print(autos['unrepaired_damage'].unique())

Index(['date_crawled', 'name', 'price', 'ab_test', 'vehicle_type',
       'registration_year', 'gearbox', 'power_ps', 'model', 'odometer_km',
       'registration_month', 'fuel_type', 'brand', 'unrepaired_damage',
       'ad_created', 'postal_code', 'last_seen'],
      dtype='object')
['manual' 'automatic' nan]
['no' nan 'yes']


# Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321.

In [35]:
#strip[0], replace-.astype(int)
#date_crawled,ad_created,last_seen
autos.describe(include='all')
def date_to_num(date):
    autos[date] = (autos[date]
                   .str.split()
                   .str[0]
                   .str.replace('-','')
                   .astype(int)
                  ) 

'''
autos['date_crawled'] = (autos['date_crawled']
                         .str.split()
                         .str[0]
                         .str.replace('-','')
                         .astype(int)
                        )
'''


"\nautos['date_crawled'] = (autos['date_crawled']\n                         .str.split()\n                         .str[0]\n                         .str.replace('-','')\n                         .astype(int)\n                        )\n"

In [36]:
date_to_num('date_crawled')
date_to_num('ad_created')
date_to_num('last_seen')
print(autos['date_crawled'].head())
print(autos['ad_created'].head())
print(autos['last_seen'].head())

0    20160326
1    20160404
2    20160326
3    20160312
4    20160401
Name: date_crawled, dtype: int64
0    20160326
1    20160404
2    20160326
3    20160312
4    20160401
Name: ad_created, dtype: int64
0    20160406
1    20160406
2    20160406
3    20160315
4    20160401
Name: last_seen, dtype: int64


In [37]:
#autos['date_crawled'].value_counts(normalize=True)
autos['date_crawled'].value_counts().sort_values(ascending=False).head(20)

20160403    1810
20160320    1775
20160321    1742
20160312    1719
20160404    1709
20160314    1696
20160307    1692
20160402    1660
20160319    1618
20160328    1615
20160315    1604
20160329    1592
20160330    1578
20160401    1578
20160308    1566
20160309    1552
20160322    1533
20160311    1515
20160310    1505
20160323    1503
Name: date_crawled, dtype: int64

In [38]:
autos['last_seen'].value_counts().sort_values(ascending=False).head(20)

20160406    10425
20160407     6197
20160405     5854
20160317     1311
20160403     1174
20160402     1151
20160330     1149
20160404     1126
20160312     1109
20160331     1103
20160401     1071
20160329     1031
20160322      973
20160328      966
20160320      963
20160321      961
20160324      919
20160325      884
20160323      857
20160326      784
Name: last_seen, dtype: int64

# Find the most common brand/model combinations

In [39]:
model_counts = autos['model'].value_counts(normalize=True).head(30)
print(model_counts)
common_brands = model_counts[model_counts > 0.01].index
common_brands

golf           0.083326
andere         0.075818
3er            0.058780
polo           0.036167
corsa          0.035785
passat         0.030323
astra          0.030300
a4             0.027670
c_klasse       0.025535
5er            0.025445
e_klasse       0.021534
a3             0.018544
a6             0.017915
focus          0.017128
fiesta         0.016229
transporter    0.015150
twingo         0.013824
2_reihe        0.013487
fortwo         0.012363
vectra         0.012228
a_klasse       0.012116
1er            0.011711
3_reihe        0.010924
mondeo         0.010767
clio           0.010632
touran         0.009733
punto          0.009328
zafira         0.008856
ka             0.007845
megane         0.007530
Name: model, dtype: float64


Index(['golf', 'andere', '3er', 'polo', 'corsa', 'passat', 'astra', 'a4',
       'c_klasse', '5er', 'e_klasse', 'a3', 'a6', 'focus', 'fiesta',
       'transporter', 'twingo', '2_reihe', 'fortwo', 'vectra', 'a_klasse',
       '1er', '3_reihe', 'mondeo', 'clio'],
      dtype='object')

In [40]:
autos.head()

date_crawled                                               name  price  \
0      20160326                   Peugeot_807_160_NAVTECH_ON_BOARD   5000   
1      20160404         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   8500   
2      20160326                         Volkswagen_Golf_1.6_United   8990   
3      20160312  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   4350   
4      20160401  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   1350   

   ab_test vehicle_type  registration_year    gearbox  power_ps   model  \
0  control          bus               2004     manual       158  andere   
1  control    limousine               1997  automatic       286     7er   
2     test    limousine               2009     manual       102    golf   
3  control   kleinwagen               2007  automatic        71  fortwo   
4     test        kombi               2003     manual         0   focus   

   odometer_km  registration_month fuel_type       brand unrepaired_damage  \
0       150000                   3       lpg     peugeot                no   
1       150000                   6    benzin         bmw                no   
2        70000                   7    benzin  volkswagen                no   
3        70000                   6    benzin       smart                no   
4       150000                   7    benzin        ford                no   

   ad_created  postal_code  last_seen  
0    20160326        79588   20160406  
1    20160404        71034   20160406  
2    20160326        35394   20160406  
3    20160312        33729   20160315  
4    20160401        39218   20160401

In [41]:
common_brandmodel = autos[['brand','model']]
common_brandmodel

brand     model
0             peugeot    andere
1                 bmw       7er
2          volkswagen      golf
3               smart    fortwo
4                ford     focus
5            chrysler   voyager
6          volkswagen      golf
7          volkswagen      golf
8                seat     arosa
9             renault    megane
11      mercedes_benz       NaN
12              smart    fortwo
13               audi        a3
14            renault      clio
15     sonstige_autos       NaN
16               opel    vectra
17         volkswagen  scirocco
18                bmw       3er
19              mazda    andere
20               audi        a4
21            porsche       911
22               mini    cooper
23            peugeot       NaN
24                bmw       5er
25               ford       NaN
26         volkswagen      polo
28               mini    cooper
29      mercedes_benz  e_klasse
30            peugeot   2_reihe
31      mercedes_benz  c_klasse
...               ...       ...
49968   mercedes_benz    andere
49969          nissan   x_trail
49970         citroen        c4
49971      volkswagen      lupo
49972   mercedes_benz      vito
49973   mercedes_benz       slk
49975           honda      jazz
49976            audi        80
49977   mercedes_benz  c_klasse
49978   mercedes_benz  e_klasse
49979      volkswagen      polo
49980            ford    escort
49981            opel     astra
49982           skoda     fabia
49983            ford     focus
49985      volkswagen       NaN
49986        chrysler      300c
49987            audi        a3
49988             bmw       3er
49989      volkswagen      polo
49990   mercedes_benz  a_klasse
49991         renault    twingo
49992            fiat    andere
49993            audi       NaN
49994            audi        a6
49995            audi        q5
49996            opel     astra
49997            fiat       500
49998            audi        a3
49999            opel    vectra

[46681 rows x 2 columns]

In [50]:
'''
brand_mean_mileage = {}
for brand in common_brands:
    brand_only = autos[autos['brand'] == brand]
    mean_mileage = brand_only['odometer_km'].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

bmm_series = pd.Series(brand_mean_mileage)
pd.DataFrame(bmm_series,columns=['mean_mileage'])
'''
models = autos['model'].unique()
#top_brand_model = []
for brand in common_brands:
    selected_rows = autos[autos['brand'] == brand]
    for model in models:
        sorted_rows = selected_rows.sort_values('model', ascending=False)
        common_bm = sorted_rows[model].value_counts(normalize=True, ascending=False).head()
common_bm

KeyError: 'andere'